Package crash on start

Crash in version `1.4.0`

This error is thrown when running `bulk-npm-publish`
```
internal/modules/cjs/loader.js:895
    throw err;
    ^

Error: Cannot find module '../dist/index.js'
Require stack:
- C:\Users\rluva\AppData\Roaming\nvm\v13.1.0\node_modules\bulk-npm-publish\bin\bulk-npm-publish.js
←[90m    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:892:15)←[39m
←[90m    at Function.Module._load (internal/modules/cjs/loader.js:785:27)←[39m
←[90m    at Module.require (internal/modules/cjs/loader.js:956:19)←[39m
←[90m    at require (internal/modules/cjs/helpers.js:74:18)←[39m
    at Object.<anonymous> (C:\Users\rluva\AppData\Roaming\nvm\v13.1.0\node_modules\←[4mbulk-npm-publish←[24m\bin\bulk-npm-publish.js:6:1)
←[90m    at Module._compile (internal/modules/cjs/loader.js:1063:30)←[39m
←[90m    at Object.Module._extensions..js (internal/modules/cjs/loader.js:1103:10)←[39m
←[90m    at Module.load (internal/modules/cjs/loader.js:914:32)←[39m
←[90m    at Function.Module._load (internal/modules/cjs/loader.js:822:14)←[39m
←[90m    at Function.Module.runMain (internal/modules/cjs/loader.js:1143:12)←[39m {
  code: ←[32m'MODULE_NOT_FOUND'←[39m,
  requireStack: [
    ←[32m'C:\\Users\\rluva\\AppData\\Roaming\\nvm\\v13.1.0\\node_modules\\bulk-npm-publish\\bin\\bulk-npm-publish.js'←[39m
  ]
}
```

--- a/.gitignore
+++ b/.gitignore
@@ -3,3 +3,4 @@ node_modules/
 .env
 !files-for-tests/.env
 logs/
+lib/

--- a/.travis.yml
+++ b/.travis.yml
@@ -20,5 +20,5 @@ jobs:
           node_js: 'node' # pre-installed version
           script:
               - npm install -g semantic-release@^15
-              - npm run compile
+              - npm run build
               - semantic-release

--- a/bin/bulk-npm-publish.js
+++ b/bin/bulk-npm-publish.js
@@ -3,4 +3,4 @@
 // (otherwise it will be executed as a shell script (hence the errors).) @see https://github.com/rluvaton/bulk-npm-publish/issues/33#issuecomment-596917110
 // @see https://stackoverflow.com/a/34354713
 
-require('../dist/index.js');
+require('../lib/index.js');

--- a/jest.config.js
+++ b/jest.config.js
@@ -3,6 +3,6 @@ module.exports = {
     testEnvironment: "node",
     verbose: true,
     testMatch: [
-        "<rootDir>/**/*.spec.ts",
+        "<rootDir>/tests/**/*.spec.ts",
     ]
 };

--- a/package.json
+++ b/package.json
@@ -5,12 +5,20 @@
   "main": "src/index.ts",
   "scripts": {
     "postinstall": "patch-package",
-    "pre:compile": "rm -rf ./dist",
-    "compile": "npm run pre:compile && tsc",
+    "prebuild": "rm -rf ./lib",
     "start": "ts-node src/index.ts --no-cache",
     "start:prod": "node ./bin/bulk-npm-publish.js",
     "start:debug": "nodemon %NODE_DEBUG_OPTION% --config nodemon-debug.json",
-    "test": "jest",
+
+    "test": "npm run lint && npm run test:unit",
+    "test:unit": "jest .",
+    "lint": "tslint -p . --fix",
+
+    "build": "npm run clean:some && tsc -p .",
+    "global": "npm i -g && bulk-npm-publish",
+    "clean:some": "rm -rf ./lib",
+    "clean:all": "rm -rf ./node_modules ./package-lock.json ./lib",
+    "refresh": "npm run clean:all && npm install",
     "semantic-release": "semantic-release"
   },
   "husky": {
@@ -22,7 +30,7 @@
     "bulk-npm-publish": "./bin/bulk-npm-publish.js"
   },
   "files": [
-    "dist/**/*"
+    "lib/**/*"
   ],
   "repository": {
     "type": "git",

--- a/src/file-writer.ts
+++ b/src/file-writer.ts
@@ -8,7 +8,7 @@ import {PathLike, WriteFileOptions} from 'fs';
  */
 const writeFile = (path: PathLike | number, data: any, options?: WriteFileOptions): Promise<void> => {
   return new Promise((resolve, reject) => {
-    fs.writeFile(path, data, options, (err) => {
+    fs.writeFile(path, data, options ?? {}, (err) => {
       if (err) {
         return reject(err);
       }

--- a/src/index.ts
+++ b/src/index.ts
@@ -47,6 +47,10 @@ const run = async () => {
   logger.verbose(`Scan complete, found ${packages.length} packages`);
 
   if (config.onlyNew.enable) {
+    if (!config.onlyNew.currentStoragePath) {
+      logger.error('Current storage is undefined');
+      return;
+    }
     logger.info(bold().underline(`Scanning Exist Storage ${emoji.get(':card_file_box:')}`));
     logger.verbose(`Scanning for exist packages in the provided storage path (${config.onlyNew.currentStoragePath})`);
     const existPackages: Package[] = storageExplorer(config.onlyNew.currentStoragePath);
@@ -63,7 +67,7 @@ const run = async () => {
 
   logger.info(bold().underline(`Creating Script ${emoji.get(':pencil2:')}`));
   logger.verbose(`Creating publish script with this options`, {options: config.npmPublishOptions});
-  let scripts: string[] = npmPublishScriptCreator(packages, config.npmPublishOptions);
+  const scripts: string[] = npmPublishScriptCreator(packages, config.npmPublishOptions);
   logger.verbose(`Script creating finished`);
 
   logger.info(bold().underline(`Writing script file ${emoji.get(':pencil:')}`));
@@ -78,6 +82,6 @@ const run = async () => {
   logger.verbose(`Finish writing script`);
 
   logger.info(bold().underline(emoji.emojify(':fire: Finish! :fire:')));
-}
+};
 
 run();

--- a/src/npm-publish-script-creator.ts
+++ b/src/npm-publish-script-creator.ts
@@ -17,7 +17,7 @@ const npmPublishScriptCreator = (
     // Combine the options
     .join(' ');
 
-  // We trim only right because there will only be a space in case of empty script options 
+  // We trim only right because there will only be a space in case of empty script options
   return packages.map(({ path }) => `npm publish ${path} ${scriptOptions}`.trimRight());
 };
 

--- a/src/storage-explorer.ts
+++ b/src/storage-explorer.ts
@@ -83,7 +83,7 @@ function getVersionFromFileName(packageName, fullPackageName: string): string {
 }
 
 const storageExplorer = (dir: string): Package[] => {
-  const packages: Package[] = [];
+  let packages: Package[] = [];
   let packageVersion: string;
 
   // Get all folder in dir
@@ -97,16 +97,16 @@ const storageExplorer = (dir: string): Package[] => {
 
   const storageFolders = filteredTree.children;
 
-  storageFolders.forEach((packagesFolders) => {
+  storageFolders?.forEach((packagesFolders) => {
     const packageOrScopeName: string = packagesFolders.name;
 
-    packagesFolders.children.forEach((packageOrScope) => {
+    packagesFolders?.children?.forEach((packageOrScope) => {
       switch (packageOrScope.type) {
         case 'directory':
           // Meaning it's a scope
           const packageName = packageOrScope.name;
 
-          const scopePackages = packageOrScope.children
+          const scopePackages = (packageOrScope?.children || [])
             // Fix #9 (adding none `tgz` file to the script)
             .filter((packageInScope) => packageInScope.type === 'file' && packageInScope.extension === '.tgz')
             .map((packageInScope) => {
@@ -126,7 +126,7 @@ const storageExplorer = (dir: string): Package[] => {
             });
 
           // Can't `concat` because packages is `constant`
-          packages.push(...scopePackages);
+          packages = packages.concat(...scopePackages);
           break;
         case 'file':
           // without scope

--- a/src/user-option/interactive/user-option-prompt-getter.ts
+++ b/src/user-option/interactive/user-option-prompt-getter.ts
@@ -21,19 +21,19 @@ const _questions = [
     type: 'text',
     name: 'npmPublishOptions.registry',
     message: `What is the registry url you want to publish to`,
-    initial: DEFAULT_USER_OPTIONS.npmPublishOptions.registry,
+    initial: DEFAULT_USER_OPTIONS?.npmPublishOptions?.registry,
   },
   {
     type: 'confirm',
     name: 'createNewOnly',
     message: `Should publish only new packages`,
-    initial: DEFAULT_USER_OPTIONS.onlyNew.enable,
+    initial: DEFAULT_USER_OPTIONS?.onlyNew?.enable,
   },
   {
     type: (createNewOnlyRes) => createNewOnlyRes ? 'text' : null,
     name: 'currentStorage',
     message: `What's the path for the current storage`,
-    initial: DEFAULT_USER_OPTIONS.onlyNew.currentStoragePath,
+    initial: DEFAULT_USER_OPTIONS?.onlyNew?.currentStoragePath,
     validate: (path) => !!path
   },
 ];

--- a/src/user-option/user-options-getter.ts
+++ b/src/user-option/user-options-getter.ts
@@ -10,8 +10,8 @@ import {logger} from '../logger';
  * @throws Error Reject if the couldn't get any options
  */
 export const userOptionGetter: (userOptionGetters: IUserOptionGetter[]) => ReturnType<IUserOptionGetter> = async (userOptionGetters: IUserOptionGetter[]) => {
-  if(!userOptionGetters || userOptionGetters.length === 0) {
-    throw new Error('userOptionGetters not provided or has no items in it')
+  if (!userOptionGetters || userOptionGetters.length === 0) {
+    throw new Error('userOptionGetters not provided or has no items in it');
   }
 
   let options: UserOptions;
@@ -22,6 +22,11 @@ export const userOptionGetter: (userOptionGetters: IUserOptionGetter[]) => Retur
       options = setDefaultUserOptionsProperties(options, DEFAULT_USER_OPTIONS);
     } catch (e) {
       logger.debug('Couldn\'t get userOptions, thrown error:', e);
+      if (e.message === 'Cancelled') {
+        logger.debug('Cancelled', e);
+        return Promise.reject(new Error('cancel'));
+
+      }
       continue;
     }
 

--- a/src/user-option/user-options.ts
+++ b/src/user-option/user-options.ts
@@ -2,20 +2,20 @@ import { NpmPublishOptions } from '../npm-publish-script-creator';
 import { getCurrentOS, OSTypes } from '../utils';
 
 export interface UserOptionsGetNewPackages {
-  enable?: boolean,
-  currentStoragePath?: string
+  enable?: boolean;
+  currentStoragePath?: string;
 }
 
 export interface UserOptions {
   storagePath: string;
-  destPublishScriptFilePath?: string;
-  npmPublishOptions?: NpmPublishOptions;
-  onlyNew?: UserOptionsGetNewPackages
+  destPublishScriptFilePath: string;
+  npmPublishOptions: NpmPublishOptions;
+  onlyNew: UserOptionsGetNewPackages;
 }
 
 const getExtensionBasedOnOS = () => {
   const os = getCurrentOS();
-  
+
   switch (os) {
     case OSTypes.LINUX:
       return 'sh';
@@ -23,9 +23,10 @@ const getExtensionBasedOnOS = () => {
     default:
       return 'bat';
   }
-}
+};
 
 export const DEFAULT_USER_OPTIONS: UserOptions = {
+  // @ts-ignore:disable-next-line
   storagePath: undefined,
   destPublishScriptFilePath: `./publish.${getExtensionBasedOnOS()}`,
   npmPublishOptions: {

--- a/src/utils.ts
+++ b/src/utils.ts
@@ -1,5 +1,5 @@
 import * as lodashDeepClone from 'lodash.clonedeep';
-import { platform } from 'process';
+import {platform} from 'process';
 
 export function deepClone<T = any>(val: T): T {
   return lodashDeepClone(val);
@@ -10,7 +10,7 @@ export enum OSTypes {
   WINDOWS = 'windows'
 }
 
-export const getCurrentOS = (): OSTypes => {
+export const getCurrentOS = (): OSTypes | undefined => {
   switch (platform) {
     case 'win32':
       return OSTypes.WINDOWS;
@@ -24,4 +24,4 @@ export const getCurrentOS = (): OSTypes => {
     default:
       return undefined;
   }
-}
+};

--- a/tests/unit/file-writer.spec.ts
+++ b/tests/unit/file-writer.spec.ts
@@ -1,6 +1,6 @@
 import 'jest-extended';
 import * as mock from 'mock-fs';
-import fileWriter from './file-writer';
+import fileWriter from '../../src/file-writer';
 
 describe('File Writer', () => {
   afterEach(() => {

--- a/tests/unit/npm-publish-script-creator.spec.ts
+++ b/tests/unit/npm-publish-script-creator.spec.ts
@@ -1,6 +1,6 @@
 import 'jest-extended';
-import npmPublishScriptCreator from './npm-publish-script-creator';
-import { Package } from './storage-explorer';
+import npmPublishScriptCreator from '../../src/npm-publish-script-creator';
+import { Package } from '../../src/storage-explorer';
 
 describe('NPM Publish Script Creator', () => {
   it('should be defined', () => {

--- a/tests/unit/storage-explorer.spec.ts
+++ b/tests/unit/storage-explorer.spec.ts
@@ -1,6 +1,6 @@
 import 'jest-extended';
 import * as mock from 'mock-fs';
-import storageExplorer, {Package} from './storage-explorer';
+import storageExplorer, {Package} from '../../src/storage-explorer';
 
 describe('Storage Explorer', () => {
 

--- a/tests/unit/user-option/env/user-option-env-getter.spec.ts
+++ b/tests/unit/user-option/env/user-option-env-getter.spec.ts
@@ -1,5 +1,5 @@
 import 'jest-extended';
-import {setPlatform} from '../../../tests/util';
+import {setPlatform} from '../../../util';
 // Doing these because we drop support to env files
 setPlatform('linux');
 
@@ -9,12 +9,12 @@ let UserOptionEnvGetter: any;
 import * as mock from 'mock-fs';
 import * as dotenv from 'dotenv';
 import {DotenvConfigOutput} from 'dotenv';
-import {IUserOptionGetter} from '../i-user-option-getter';
+import {IUserOptionGetter} from '../../../../src/user-option/i-user-option-getter';
 import createSpy = jasmine.createSpy;
 import Spy = jasmine.Spy;
 import Mock = jest.Mock;
 
-const setEnv = (dotEnvFileContent: string) => {
+const setEnv = (dotEnvFileContent?: string) => {
   if (dotEnvFileContent) {
     mock({
       envFiles: {
@@ -37,8 +37,8 @@ describe('Get User Options from Environment file', () => {
   beforeAll(() => {
     jest.isolateModules(() => {
       // const utils = require('../utils');
-      UserOptions = require('../user-options').UserOptions;
-      UserOptionEnvGetter = require('./user-option-env-getter').userOptionEnvGetter;
+      UserOptions = require('../../../../src/user-option/user-options').UserOptions;
+      UserOptionEnvGetter = require('../../../../src/user-option/env/user-option-env-getter').userOptionEnvGetter;
       userOptionEnvGetter = jest.fn(UserOptionEnvGetter);
     });
   });

--- a/tests/unit/user-option/interactive/user-option-prompt-getter.spec.ts
+++ b/tests/unit/user-option/interactive/user-option-prompt-getter.spec.ts
@@ -2,16 +2,16 @@ import 'jest-extended';
 
 import createSpy = jasmine.createSpy;
 import Spy = jasmine.Spy;
-import {IUserOptionGetter as IUserOptionGetterLib} from '../i-user-option-getter';
-import {UserOptions as UserOptionsLib} from '../user-options';
-import {setPlatform} from '../../../tests/util';
+import {IUserOptionGetter as IUserOptionGetterLib} from '../../../../src/user-option/i-user-option-getter';
+import {UserOptions as UserOptionsLib} from '../../../../src/user-option/user-options';
+import {setPlatform} from '../../../util';
 
 
 function getDeps(): { prompts: any, UserOptions: UserOptionsLib, IUserOptionGetter: IUserOptionGetterLib, UserOptionPromptGetter: IUserOptionGetterLib } {
   const prompts = require('prompts');
-  const {UserOptions} = require('../user-options');
-  const {IUserOptionGetter} = require('../i-user-option-getter');
-  const {userOptionPromptGetter: UserOptionPromptGetter} = require('./user-option-prompt-getter');
+  const {UserOptions} = require('../../../../src/user-option/user-options');
+  const {IUserOptionGetter} = require('../../../../src/user-option/i-user-option-getter');
+  const {userOptionPromptGetter: UserOptionPromptGetter} = require('../../../../src/user-option/interactive/user-option-prompt-getter');
 
   return {prompts, UserOptions, IUserOptionGetter, UserOptionPromptGetter};
 }
@@ -54,7 +54,7 @@ describe('Get User Options from User Input', () => {
   });
 
   it('userOptionPromptGetter should be define', () => {
-    const {userOptionPromptGetter} = require('./user-option-prompt-getter');
+    const {userOptionPromptGetter} = require('../../../../src/user-option/interactive/user-option-prompt-getter');
     expect(userOptionPromptGetter).toBeDefined();
   });
 
@@ -131,7 +131,7 @@ describe('Get User Options from User Input', () => {
     };
 
     // Inject the values
-    prompts.inject([expectedUserOptions.storagePath, expectedUserOptions.destPublishScriptFilePath, expectedUserOptions.npmPublishOptions.registry]);
+    prompts.inject([expectedUserOptions.storagePath, expectedUserOptions.destPublishScriptFilePath, expectedUserOptions?.npmPublishOptions?.registry]);
     await testUserOptionPromptGetter<typeof UserOptions>(userOptionPromptGetter, expectedUserOptions);
   });
 
@@ -153,9 +153,9 @@ describe('Get User Options from User Input', () => {
     prompts.inject([
       expectedUserOptions.storagePath,
       expectedUserOptions.destPublishScriptFilePath,
-      expectedUserOptions.npmPublishOptions.registry,
-      expectedUserOptions.onlyNew.enable,
-      expectedUserOptions.onlyNew.currentStoragePath
+      expectedUserOptions?.npmPublishOptions?.registry,
+      expectedUserOptions?.onlyNew?.enable,
+      expectedUserOptions?.onlyNew?.currentStoragePath
     ]);
 
     await testUserOptionPromptGetter<typeof UserOptions>(userOptionPromptGetter, expectedUserOptions);

--- a/tests/unit/user-option/user-options-getter.spec.ts
+++ b/tests/unit/user-option/user-options-getter.spec.ts
@@ -1,16 +1,16 @@
 import 'jest-extended';
-import {UserOptions as UserOptionsLib} from './user-options';
+import {UserOptions, UserOptions as UserOptionsLib} from '../../../src/user-option/user-options';
 
-import {IUserOptionGetter as IUserOptionGetterLib} from './i-user-option-getter';
+import {IUserOptionGetter as IUserOptionGetterLib} from '../../../src/user-option/i-user-option-getter';
 import Spy = jasmine.Spy;
 import createSpy = jasmine.createSpy;
-import {setPlatform} from '../../tests/util';
+import {setPlatform} from '../../util';
 
 
 function getDeps(): { UserOptions: UserOptionsLib, IUserOptionGetter: IUserOptionGetterLib, userOptionGetter: IUserOptionGetterLib } {
-  const {UserOptions} = require('./user-options');
-  const {IUserOptionGetter} = require('./i-user-option-getter');
-  const {userOptionGetter} = require('./user-options-getter');
+  const {UserOptions} = require('../../../src/user-option/user-options');
+  const {IUserOptionGetter} = require('../../../src/user-option/i-user-option-getter');
+  const {userOptionGetter} = require('../../../src/user-option/user-options-getter');
 
   return {UserOptions, IUserOptionGetter, userOptionGetter};
 }
@@ -102,7 +102,7 @@ describe('Get User Options (from the available option)', () => {
       }
     };
 
-    const pr = getOptionsAndEnsureCalledTimeAndArgs<typeof IUserOptionGetter>(userOptionGetter, [
+    const pr = getOptionsAndEnsureCalledTimeAndArgs<typeof IUserOptionGetter | any>(userOptionGetter, [
       () => Promise.resolve({
         storagePath: expectedUserOptions.storagePath,
         destPublishScriptFilePath: expectedUserOptions.destPublishScriptFilePath,

--- a/tests/unit/user-option/user-options.spec.ts
+++ b/tests/unit/user-option/user-options.spec.ts
@@ -1,14 +1,15 @@
 import 'jest-extended';
-import {setDefaultUserOptionsProperties, UserOptions} from './user-options';
+import {setDefaultUserOptionsProperties, UserOptions} from '../../../src/user-option/user-options';
 
 describe('UserOptions', () => {
   describe('#setDefaultUserOptionsProperties', () => {
     it('when not destPublishScriptFilePath and npmPublishOptions.registry should set this to default', () => {
-      const currOptions: UserOptions = {
+      const currOptions: UserOptions | any = {
         storagePath: 'C://',
       };
 
-      const defaultOptions: UserOptions = {
+      const defaultOptions: UserOptions | any = {
+        // @ts-ignore:disable-next-line
         storagePath: undefined,
         destPublishScriptFilePath: './publish.bat',
         npmPublishOptions: {
@@ -50,6 +51,7 @@ describe('UserOptions', () => {
       };
 
       const defaultOptions: UserOptions = {
+        // @ts-ignore:disable-next-line
         storagePath: undefined,
         destPublishScriptFilePath: './publish.bat',
         npmPublishOptions: {

--- a/tsconfig.json
+++ b/tsconfig.json
@@ -9,9 +9,16 @@
     "typeRoots": ["node_modules/@types", "typings"],
     "noImplicitAny": false,
     "sourceMap": false,
-    "outDir": "./dist"
+    "outDir": "lib",
+    "strict": true,
+    "types": ["node"],
+    "resolveJsonModule": true
   },
   "compileOnSave": false,
+  "include": [
+    "src",
+    "bin"
+  ],
   "exclude": [
     "node_modules"
   ]

